clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java

```diff
@@ -100,6 +100,7 @@ private KafkaMbean removeAttribute(KafkaMetric metric) {
         MetricName metricName = metric.metricName();
         String mBeanName = getMBeanName(prefix, metricName);
         KafkaMbean mbean = this.mbeans.get(mBeanName);
+        mbeans.remove(mBeanName);
```

Comment:
We can remove only if mbean.metrics.isEmpty. This line should be added in metricRemoval() method after unregister call.